airflow/models/__init__.py

```diff
@@ -4244,6 +4254,63 @@ def _test_cycle_helper(self, visit_map, task_id):
 
         visit_map[task_id] = DagBag.CYCLE_DONE
 
+    def _set_dag_permissions(self, access_control):
+        """Grant permissions on this DAG's ViewModel to the given role(s).
+
+        :param access_control: a dict where each key is a rolename and
+            each value is a set() of permission names (e.g.,
+            {'can_dag_read'}
+        :type access_control: dict
+
+        """
+        sm = cached_appbuilder().sm
```

Comment:
this may be a little concern as it requires starting the webserver when Airflow starts the scheduler.

Two ideas could help solve these cases:
1. move all FAB models(https://github.com/dpgaspar/Flask-AppBuilder/blob/1e900bba85452de6d988f7da191f9a26fec62226/flask_appbuilder/security/sqla/models.py) into Airflow code base(cc @mistercrunch  who brought this idea previously). This helps Airflow not relied on FAB that much. 
2. Created an API endpoint in web server for creating permission / delete permission purpose(https://issues.apache.org/jira/browse/AIRFLOW-2693). In this way, the Airflow web server handles all the FAB permission actions. And scheduler just delegates the permission handling to the webserver.

cc @jgao54  @mistercrunch

Reply:
@feng-tao Yeah, I had the same thought - I really don't like doing this in the `__init__` method.

I see that we are already doing some permission handling in the `/refresh` endpoint - maybe this would be a natural place to invoke `dag.set_dag_permissions()`?

https://github.com/apache/airflow/blob/81f7c475a92b40320a683fc0bf81b62c47052238/airflow/www/views.py#L1682

We could have similar logic in the `sync_perm` CLI command: for each DAG in the DagBag, call `dag.set_dag_permissions()`

https://github.com/apache/airflow/blob/81f7c475a92b40320a683fc0bf81b62c47052238/airflow/bin/cli.py#L1634

Reply:
I think extending the `/refresh` endpoint should work. We need to pass the `access_control` info from scheduler to webserver endpoint during the DAG `_init_` constructor.  I am not so sure how to put the `set_dag_permissions` function inside `sync_perms` cli though.

I think we couldn't get the dag access_control information until scheduler parses the dag information. And we need to persist these infomation somewhere in the db. Then the cli knows how to persist the permission with roles mapping information, right? In this case, I don't know how the CLI could figure how to call the ``sync_perm_for_dag``.

Reply:
sync offline, and @astahlman will move the logic into the `sync_perm` cli.

Reply:
Updated - I've moved this logic from the DAG class to the AirflowSecurityManager and triggered it from the CLI `sync_perm` command as well as the `/refresh` and `/refresh_all` endpoints. This way is much better I think - thanks for the feedback.